query/Inserter.java

```diff
@@ -113,18 +111,27 @@ public static Inserter create(Reasoner reasoner, ConceptManager conceptMgr, Type
                 assert !filter.isEmpty();
                 matcher = Matcher.create(reasoner, match.get(filter));
             }
-
+            VariableRegistry registry = VariableRegistry.createFromThings(query.variables());
+            for (Variable variable : registry.variables()) validate(variable, matcher);
             return new Inserter(matcher, conceptMgr, registry.things(), context);
         }
     }
 
-    public static void validate(Variable var) {
+    public static void validate(Variable var, @Nullable Matcher matcher) {
         try (FactoryTracingThreadStatic.ThreadTrace ignored = traceOnThread(TRACE_PREFIX + "validate")) {
-            if (var.isThing()) validate(var.asThing());
+            if (var.isType()) validate(var.asType(), matcher);
+            else if (var.isThing()) validate(var.asThing());
             else if (var.isValue()) validate(var.asValue());
         }
     }
 
+    private static void validate(TypeVariable var, @Nullable Matcher matcher) {
```

Comment:
Drive-by fix: we should throw a useful exception if trying to use an unbound type variable in an insert query